src/client/react.js

```diff
@@ -43,8 +43,33 @@ const logger = proxyLogger(_logger, __NEXTAUTH.basePath)
 /** @type {import("types/internals/react").SessionContext} */
 const SessionContext = React.createContext()
 
-export function useSession() {
-  return React.useContext(SessionContext)
+export function useSession(options = {}) {
+  const value = React.useContext(SessionContext)
+
+  if (process.env.NODE_ENV !== "production" && !value) {
+    throw new Error("useSession must be wrapped in a SessionProvider")
+  }
+
+  const { required, action } = options
+
+  const requiredAndNotLoading = required && value.status === "unauthenticated"
+
+  React.useEffect(() => {
+    if (requiredAndNotLoading) {
+      const url = `/api/auth/signin?${new URLSearchParams({
+        error: "SessionRequired",
+        callbackUrl: window.location.href,
+      })}`
+      if (action) action()
+      else window.location.replace(url)
+    }
+  }, [requiredAndNotLoading, action])
+
+  if (requiredAndNotLoading) {
+    return { data: value.data, loading: true, status: "loading" }
```

Comment:
I think we should have just `status: loading` and drop boolean flags, in this case `loading`. 

`data` seems a bit generic too, I would call it `session`.

Usually, boolean flags [just end creating technical debt](https://medium.com/geekculture/why-you-shouldnt-use-booleans-in-rest-apis-2747b187876c). 📐

Reply:
Yeah, I was a bit hesitant, as migration would have been simpler. But I am on board to drop the `loading` flag anyway.

I think of v4 as quite a few breaking changes but in the name of creating a good base for future improvements. Refactoring the APIs to use objects in most situations (returned values or function parameters)` will make it possible to extend almost anything with new options/fields without requiring the user to change their APIs.